Fix custom resource icons in FileSystem

KoBeWi opened the pull request:
Fixes #32706
Supersedes #73611

KoBeWi committed:
Fix custom resource icons in FileSystem

KoBeWi commented on editor/filesystem_dock.cpp:
```diff
@@ -256,10 +263,11 @@ bool FileSystemDock::_create_tree(TreeItem *p_parent, EditorFileSystemDirectory
-			file_item->set_icon(0, _get_tree_item_icon(!fi.import_broken, fi.type));
-			String file_metadata = lpath.path_join(fi.name);
+			file_item->set_icon(0, _get_tree_item_icon(!fi.import_broken, file_metadata, fi.type));
+			file_item->set_icon_max_width(0, 16 * EDSCALE);
```
I think instead of hard-coding 16 we could use width of an existing icon, but for that it would be better to use some cached one (and there is no theme cache yet). I could leave a TODO comment maybe? (or add theme cache I guess)

YuriSizov replied:
You should use `get_theme_constant(SNAME("class_icon_size"), SNAME("Editor"))` instead.

YuriSizov commented on editor/filesystem_dock.cpp:
```diff
@@ -254,12 +261,15 @@ bool FileSystemDock::_create_tree(TreeItem *p_parent, EditorFileSystemDirectory
 		_sort_file_info_list(file_list);
 
 		// Build the tree.
+		int icon_size = get_theme_constant(SNAME("class_icon_size"), SNAME("Editor"));
```
Can probably be a const variable.

YuriSizov commented:
Yeah, the branch name is very fitting 🙃 This approach is going to read and parse the file for each stored resources every time we need to figure out its icon. Even if this information doesn't change, we would still parse the file, at least partially, to fetch the type information.

This definitely needs some cache, and that's what reduz wanted as well, I believe. To utilize the editor file system somehow, and only update the information when the underlying data, like the type or the script, changes. This way we would only have to run this expensive operation on change rather than on fetch.

I believe that the PR this one is supposed to supersede was going this route. After a lot of discussion they've hooked into the preview generator process, since it already needs to run for a changed resource. Maybe it's not the best place to hook into, but at least it creates an accessible cache for us to use on the file tree refreshes.

So in the end, I would not merge this in its current state and look for a more optimal approach using some sort of cache.

KoBeWi commented:
I think the icons can be cached locally, we already do this with scene tree (see #60789). I'll look into that, I thought `get_class_icon()` already uses it...

KoBeWi commented:
Ok rebased and updated. I used EditorFileSystem's cache to get resource icon, but it has caveats:
- it assumes the first dependency in the file is its Script
- it checks whether that dependency is actually a Script (not sure how fast is that)

I added a temporary HashMap that stores cache for each script path, so it shouldn't be *too slow*, but testing appreciated.

adamscott commented on editor/filesystem_dock.cpp:
```diff
@@ -173,14 +173,48 @@ FileSystemList::FileSystemList() {
+		return String();
+	}
+
+	const String &script_path = deps[0]; // Assuming the first dependency is a script.
+	if (script_path.is_empty() || !ClassDB::is_parent_class(ResourceLoader::get_resource_type(script_path), SNAME("Script"))) {
+		return String();
+	}
```
Maybe loop through the dependencies for the first script?

KoBeWi replied:
That would be slower though 🤔
I'd change this only if it proves unreliable.

adamscott reviewed (approved):
Other than my comment, everything seems fine! I tested it and it works.

Thanks @KoBeWi !

akien-mga commented:
Thanks!

z4gon commented:
@KoBeWi I'm seeing this issue in this build:
```
Godot Engine v4.3.beta1.official.a4f2ea91a - https://godotengine.org
OpenGL API 4.1 Metal - 88 - Compatibility - Using Device: Apple - Apple M1 Pro
```
![Screenshot 2024-06-07 at 6 34 48 PM](https://github.com/godotengine/godot/assets/4588601/ea3f3ccf-7927-43ed-90bc-63cedd8c8d4c)
![Screenshot 2024-06-07 at 6 34 43 PM](https://github.com/godotengine/godot/assets/4588601/28d08996-6e30-44b5-9191-2345e5fc7a38)
![Screenshot 2024-06-07 at 6 34 04 PM](https://github.com/godotengine/godot/assets/4588601/9874e78f-63ae-4cff-97bd-1011b6925ddb)
